### ISSUE TITLE
Modrinth Mod with `side = "server"` not imported to Modrinth Correctly

### ISSUE THREAD

**Issue by harrismcc** [packwiz/packwiz#284]
When I add [Terralith from Modrinth](https://modrinth.com/mod/terralith) to the pack it's marked as `side = 'server'` which is also how it's classified on Modrinth. However, when I use packwiz to export the pack as a Modrinth modpack and then import it in the Modrinth client, it's missing from the modlist. Only when I change `side` in the toml to `side = "both"` does it show up in the client.

Is this a bug with the way packwiz is exporting the modpack? When I create the export from modrinth itself this issue does not happen.

**Comment by copygirl** [comment]
That sounds like a misclassification of the mod on Modrinth. Of course if you mark a mod as server-only, it will not be part of a client-side export of the pack. Or has that meaning changed since I was last making modpacks myself? Might be interesting to ask the people at Modrinth what "Client side: Unsupported" means and how they deal with it on their site / in their own launcher.

**Comment by copygirl** [comment]
It looks like Terralith is not the only mod. I've also seen Axes Are Weapons and YUNG's mods "suffer" from this.
Out of nowhere I was putting together a pack for some friends, so I had picked up packwiz and ran into this myself.

**Comment by harrismcc** [comment]
Yeah I think you're right it's misclassified, for now I'll just have to manually set the value to both but that's an interesting issue

**Comment by brawaru** [comment]
Hi there. I've noticed this issue just today's morning after another report of packwiz-broken modpack and wanted to add to this issue, in case this wasn't said already: API project sideness doesn't precisely match the modpack environment filters.

Modrinth modpack was designed to avoid the need for the so-called server packs, server modpacks (where server JAR, all the mod JARs, and bat/sh files are all downloaded and packed in to a separate non-modpack .zip file), so neither the creators nor the users have to be burdened with setting up a server and licensing (redistribution rights) of mods when the tooling can set up a server for them. A vision very close to ATLauncher server creation capability.

When we're talking about sides in Modrinth API projects, the `server` refers to any server, an integrated (single player/LAN) or a dedicated one (server JAR). However, since Modrinth modpacks are both client and server packs, the environment filters are used solely to separate client modpack installations from the dedicated server modpack installations (say, on server hostings or through tools like aforementioned ATLauncher).

This is outlined in ~~docs~~ [support portal article](https://support.modrinth.com/en/articles/8802351-modrinth-modpack-format-mrpack) [^1]:
> `server` refers to the *dedicated server*. Even though clients technically have a logical server, if something is marked server only, it should not be installed on the client.

Therefore, packwiz makes a wrong assumption of the side, and server-required and server-optional API projects should translate to `side = "both"` per packwiz definition, not `side = "server"`. I believe the only assumption that can be drawn from API project sides is that the server-unsupported client-required mods should be `side = "client"`. Whether to include mod only in the dedicated server modpack installation is currently prerogative of the modpack creator, likewise for optionality.

There have been conversations about splitting the Modrinth API definition of `server` into `integrated_server` and `dedicated_server`, which should help with assumptions here, and I believe this is being already experimented with in v3 API (which is unstable and should not be used). This, however, is going to take quite some time, so in a meanwhile I think it would be easier and quicker to change assumptions made by packwiz and avoid having creators think their packwiz-created modpacks are broken :)

If you have any questions, please let me know. Additionally, feel free to chat with us and other developers using Modrinth API in `#api-development` channel [on Modrinth Discord server](https://discord.modrinth.com/), so we can continue to make Modrinth API better for everyone.

[^1]: yeah, we really should've not put technical docs there, but alas

**Comment by copygirl** [comment]
@brawaru The mods mentioned (Terralith, Axes Are Weapons and YUNG's mods (not the API)) are marked as "Client: Unsupported" and "Server: Required" according to their mod pages. So according to the Modrinth "docs" you linked to, it should be installed on a dedicated server only, as it says the mod is "unsupported" on the client.

I think packwiz is doing the right thing, and the mod author marked their mod incorrectly.
How does the official Modrinth launcher handle this case? Differently from what their docs state?

**Comment by brawaru** [comment]
No, project authors marked their mods correctly. Server in Modrinth project sideness currently refers to the logical server, thus it means *both* the integrated server (singleplayer/LAN) and the dedicated server (server JAR). Therefore, Modrinth project sideness cannot be directly translated into the Modrinth modpack environment filters, where server is defined as the dedicated server only.

Terralith, for example, does nothing on the client side of things — it doesn't change UI, doesn't add any new blocks or textures. It only changes the server's terrain generation. Therefore it's marked correctly as client unsupported and server required.

However, since majority of modpacks are created with the single player support in mind, and single player is a logical server, Terralith should be installed on the client as well, and should translate to `side = "both"` per packwiz definition, or server-required client-required per Modrinth modpack format environment definition.

Unfortunately, Modrinth App is not the best example here either due to the oversight it marks absolutely all mods as required on both sides. The proper example would be Prism Launcher:

<table>
<tbody>
  <tr>
    <th rowspan="2">Mod name</th>
    <th colspan="2">Project sides</th>
    <th colspan="2">Modpack environment</th>
  </tr>
  <tr>
    <th>Client</th>
    <th>Server</th>
    <th>Client</th>
    <th>Server</th>
  </tr>
  <tr>
    <th>Sodium</th>
    <td>Requried</td>
    <td>Unsupported</td>
    <td>Required<br></td>
    <td>Unsupported</td>
  </tr>
  <tr>
    <th>Simple Voice Chat</th>
    <td>Required</td>
    <td>Required</td>
    <td>Required<br></td>
    <td>Required</td>
  </tr>
  <tr>
    <th>Terralith</th>
    <td>Unsupported</td>
    <td>Required</td>
    <td>Requried</td>
    <td>Required</td>
  </tr>
</tbody>
</table>

**Comment by Prospector** [comment]
to put it more simply, using the metadata currently available, in mrpack index files, all mods should be marked as client `required` unless otherwise indicated by the modpack author, and server should be set to `unsupported` if the mod is client-side only (Modrinth project indicates client `required` and server `unsupported`) but otherwise `required` unless otherwise indicated by the modpack author.

**Comment by copygirl** [comment]
Okay, yeah, that's very confusing. Apologies for not understanding this even though you explained it to me. To be fair, it really isn't clear, especially looking at the website and seeing "Client: Unsupported" and that not meaning what I thought it meant.

@brawaru already hinted at it, but I guess there's no way to mark a mod as dedicated server only at this time? Such as a backup mod or management tool, which would not make sense (or not work at all) on a client.

With a separate `logical_server` field, would it ever make sense to mark client optional but logical server required? I can understand separate optional handling for server and client, but ignoring that for a moment, how is `side = client/server/both` not enough for deciding whether a mod should be downloaded? Modrinth isn't a mod loader, and neither are its API consumers, so it shouldn't concern itself with such details. Am I missing something?

(I understand this might not be the best place to discuss, but hey, Discord is worse. :stuck_out_tongue:)

**Comment by brawaru** [comment]
> @brawaru already hinted at it, but I guess there's no way to mark a mod as dedicated server only at this time? Such as a backup mod or management tool, which would not make sense (or not work at all) on a client.

Unfortunately, no, not on Modrinth API side of things. Currently this is should be up to the modpack author to correct assumption made from API data, so if they use packwiz to create their modpack, they can simply change that mod's metadata file to say `side = "server"`. Sadly, though, I think no launcher currently provides ability to specify environment for mods when exporting a Modrinth modpack.

> With a separate `logical_server` field, would it ever make sense to mark client optional but logical server required?

Logical server is just Minecraft server, whether it's single player or dedicated server. You probably meant `integrated_server` (which means server integrated into Minecraft client, and which starts when you enter one of your single player worlds) and `dedicated_server` (server started from a server JAR).

This change would only affect Modrinth API, Modrinth modpack format will remain with two environment filters: `client` for when modpack is installed through the modded launcher, or `server` for when modpack is installed by the dedicated server deployment tool.

Thus, packwiz's `side = "client|server|both"` works perfectly and *is* enough. If not for `"optional"` value for environment filters, I think Modrinth modpack format would probably also had only `"side": "client|server|both"`.

This problem here is more about the assumption of `side` that packwiz makes based on the current v2 Modrinth API sides. The `server` field split in v3 should help with that, since you will be able to easily determine the side in modpack:

```go
clientRequired := modVersion.IntegratedServer || modVersion.Client
serverRequired := modVersion.DedicatedServer

if clientRequired && serverRequired {
  modManifest.Side = SideBoth
} else if clientRequired {
  modManifest.Side = SideClient
} else if serverRequired {
  modManifest.Side = SideServer
} else {
  // hello modrinth API??
  modManifest.Side = SideBoth // what's the worst that can happen
}
```

Once again, however, this has not yet finalised, and there are multiple competing proposals how this can be solved, so it's going to be a lengthy process.

The reason Modrinth is concerned with this data is exactly because its app, any other launchers and tools need it to accurately draw the conclusions on which side of modpack mod goes based on its sideness. Users find this data useful too, e.g. some people prefer to make server-only modpacks, that are 100% compatible with vanilla clients, others might want to have a modpack that only has client mods and 100% compatible with vanilla servers.

**Comment by lucascoelho07** [comment]
Any news on this? Modrinth clearly differentiates "Client **and** Server" to "Client **or** Server" in their website. Idk about the API tho. But **if possible**, it would be really good to have this working as some of us expected or at least an option to export the whole pack and not only the ones required on the client.

**Comment by offbeat-stuff** [comment]
Modrinth metadata has updated,
Packwiz should also be updated to support that
I mark my mod as supported on both dedicated and integrated server
But gets marked as server side in packwiz
